### DISCUSSION TITLE
자기장 정책

### DISCUSSION THREAD

**Discussion by tnwlsgk1002** [boostcampwm-2022/android11-almost-there#69] [Ideas]
- 처음 자기장은 어떻게 정할까?
- 자기장을 어떤 식으로 좁힐까?
- 그동안의 [작당 모의](https://www.figma.com/file/sMzBacITSUGSpR7RC5bcpt/%EC%9E%91%EB%8B%B9%EB%AA%A8%EC%9D%98?node-id=0%3A1&t=ppLYlTI8m6NacVIJ-0)

**Comment by soulplay95** [comment]
### Flow

1. 자기장 시작 시간 5분 전에 모든 참가자들에게 "5분 뒤에 게임이 시작됩니다. GPS를 키고 게임에 참여하세요!" 라는 메시지를 담은 push 알림을 보낸다.
2. 자기장 시작 시간까지 GPS를 키지 않은 참여자는 약속에서 제외한다.
3. 자기장 시작 시간이 되면, 모든 사용자의 **현재 위치(위경도)**와 **현재 유저의 이동 수단 정보**를 받아온다.
4. 약속 장소까지 해당 이동 수단으로 이동했을 때 1분(자기장 갱신 주기) 후 예상 위치를 구한다.
5. 다음 자기장은 1분 후 모든 예상 위치를 포함하는 최소 원으로 그린다.
6. 1분마다 4~5번 과정을 반복하며 자기장을 그린다.

### 최초 자기장은 어떻게 그릴까?

=> 최초 사용자들의 현재 위치를 모두 포함하는 가장 작은 원

### 자기장 갱신 주기마다 어떻게 새로 그릴까?

=> 위의 Flow대로

### 참고

- [Activity Recognition Transition API](https://developer.android.com/guide/topics/location/transitions)
- [모든 점을 포함하는 가장 작은 원 - 최소 외접원](https://greimul.tistory.com/32)

**Reply by tnwlsgk1002** [reply]
현재 유저의 이동 수단 정보를 처음에 받게 하는 건 좋은 것 같아요!
최초 자기장 위치는 현재 위치와 목적지 위치를 모두 포함하게끔 그리는 것도 좋아요!
다만 4~5번의 과정을 1분마다 반복하는 게 조금 빡세보입니다...
서버에서 일을 할 수 없으니 로컬에서 해야 할 텐데,
자기장을 처음부터 fix하지 않는다는 문제가 조금 걸리네욥

**Reply by swjw13** [reply]
1분 후의 위치를 알 수 있는 방법이 있으면 좋은 방법인 것 같습니다! 다만 예측과 실제가 달랐을 때 어떻게 할 지 정하는 과정도 빡셀 것 같네요...

**Comment by fbghgus123** [comment]
# 게임 준비

1. 자기장 시작 5분 전 모든 참가자들에게 **게임 준비 알람**을 표시한다.
2. 참여자가 게임 준비 알람을 터치하면 해당 약속의 참여 정보 DB에 해당 참여자의 초기 위치 정보를 수집 한다.
3. 수집한 초기 위치를 토대로 API 를 통해 현재 위치로부터 도착 지점까지 예상 소요 시간을 받아온다.

![image](https://user-images.githubusercontent.com/74997185/204089515-36698c6a-ecde-46a5-bba2-cfb2269ce9ed.png)

4. 참여자의 초기 위치와 도착 지점까지 일직선을 그어 예상 소요 시간 만큼 구간을 나눈다.

![image](https://user-images.githubusercontent.com/74997185/204089649-917d6c87-d734-44c6-995e-da051c82df0f.png)

6. 게임 시간 만큼 직선 거리를 확장시킨다. 

![image](https://user-images.githubusercontent.com/74997185/204091366-e3294a03-2a97-43d1-a14b-c92a53967c4e.png)

7. 각 시간 별 예상 위치를 파이어스토어에 저장한다.


# 게임 시작

1. 게임이 시작되면 포그라운드 서비스가 실행 되며, 일정 시간이 될 때 마다 참여자들의 시간별 예상 위치를 불러 온다.

2. 불러온 해당 시간의 모든 예상 위치를 모두 포함하는 외접원을 그린다.

![image](https://user-images.githubusercontent.com/74997185/204091921-07d44744-0fa8-4953-ba2d-a90cfd5412c5.png)

3. 해당 외접원의 중심과 반지름을 토대로 해당 플레이어가 자기장 안에 있는지 밖에 있는지 판단하고, HP를 깎는다.

**Reply by tnwlsgk1002** [reply]
앱의 목적인 **늑장 부리냐**에 집중할 수 있어서 좋은 것 같아요! 
다만 **현재 위치로부터 도착 지점까지 예상 소요 시간을 받아온다**에서 그 **소요 시간을 어떻게 구하느냐**가 걸리네요.

**Reply by tnwlsgk1002** [reply]
만약 대중교통으로만 한다면 다음과 같은 상황이 발생합니다..
ODSay API 저희 집에서 바로 옆에 있는 초등학교로 테스트한 결과입니다.
![image](https://user-images.githubusercontent.com/76458724/204124260-f5048c5c-3db1-45c4-adbd-f3fa8feb7119.png)

**Comment by swjw13** [comment]
![KakaoTalk_Photo_2022-11-27-01-42-38](https://user-images.githubusercontent.com/75927312/204099466-d617512b-0e01-426c-a753-53022b3e5f14.jpeg)

# 사용자의 위치에서 자동차 / 대중교통 / 도보로 목적지까지 가는 최단 시간을 이용하는 방법

## 적용 방법
1. 게임 시작시에 현재 위치에서 목적지까지 자동차 / 대중교통 / 도보로 이동 가능한 최단 시간을 구한다.
2. 1번 방법으로 구한 경로에서 시간에 대한 이동거리를 약속 시간에 맞춰서 거리를 구한다.
  - 예를 들어 3km를 30분만에 갈 수 있다하면 60분 기준으로 6km가 된다.
3. 각 유저마다 자동차 / 대중교통 / 도보로 구한 약속시간 기준 거리 중 최대값을 자기장 반지름의 후보로 설정한다.
4. 모든 유저의 반지름 후보 중 가장 긴 값을 반지름, 약속 장소를 중심 점으로 하는 원을 자기장으로 설정한다.
5. 만약 대중교통 혹은 자동차 경로가 나오지 않는다면 고려 대상에서 제외한다.
  - 예를 들어 대중교통 경로가 나오지 않는다면 자가용 / 도보 기준으로만 반지름 후보를 구한다.

## 게임 진행
1. 자기장의 크기는 일정한 비율대로 반지름이 줄어들고, 결과적으로는 약속 시간이 됐을 때 약속 장소에서 원이 소멸한다.

## 게임 시나리오
- 만약 시간보다 일찍 도착한다면 HP는 의미가 없어진다.
- 만약 시간에 맞춰 도착하더라도 경로를 잘 지킨다면 HP는 줄어들 일이 없다(길이가 같다면 곡선 위 점까지의 유클리드 거리는 직선 위 점의 유클리드 거리보다 항상 가깝기 때문)
- 아슬아슬하게 도착한다면 HP가 적게 내렸을 것이고, 많이 도착을 못 한 상태라면 HP가 많이 깎여 있는 상태일 것이다.

## 해결하지 못하는 문제점
- 제주도와 같은 특수한 지역까지는 포함할 수 없다.
- 사용하는 Map의 한계로 서울역 ~ 부산역 등 거리가 먼 거리는 자기장을 그릴 수 없다.

**Reply by tnwlsgk1002** [reply]
좋은 방법인 것 같습니다!
대중교통 / 도보 / 자가용 모두 API로 했을 때 예외 코드가 많은데
이 세 가지 API 응답을 모두 받으니까 예외 처리가 편할 것 같아요!

**Reply by tnwlsgk1002** [reply]
다만 다른 분들이 제시한 것과 마찬가지로 제주도 같은 경우는 불가능할 것 같네요...
Odsay api가 비행기는 지원해주지 않아요 ㅠㅠ
![image](https://user-images.githubusercontent.com/76458724/204129258-f4c216fe-fbb7-4719-bce8-412813ea4d8d.png)

사실 이것까지 고려하는 게 어렵긴 합니다.. :(

**Comment by tnwlsgk1002** [comment]
저는 남은 시간 내에 고려할 수 있는 것과 어려울 것 같은 걸 적어봅니다!

# 고려할 수 있는 것
## 패널티
고려되었으면 하는 건 패널티입니다.
그래도 게임이니까 각자 공평해야 하고, 고려할 수 있는 상황은 고려할 수 있었으면 좋겠어요.

## 다음과 같은 상황 고려
- 모든 점이 한쪽으로 몰려있을 때
    - 반드시 목적지를 포함 시켜야 한다.
- 만약 약속 참여 OR 생성 시 소요 시간을 선택할 수 있게 된다면?
    - 내가 그 때 어느 경로로 가게 될 지 모른다면..?

# 어려워 보이는 것
## "실제 경로"에 따른 예상 소요 시간
반지름을 잡을 때에는 **유클리드** 기준으로 원을 그리는데, 
소요 시간은 API를 사용해서 **실제 경로**를 기준으로 하고 있습니다.

저는 통일되었으면 해서 차라리 API를 사용하는 것이 아니라
모든 거리를 **유클리드 거리** 기준으로 하고, **사람 걷기 평균 속도 / 자가용 평균 속도 / 버스 평균 속도**로 하고 정해서 하는 게 좋지 않나 싶습니다.

API에 너무 의존적이고, 세 개를 동시에 처리하고 하는 게 너무 부담이 클 것 같아요. 
서버도 없는 상황에서 로컬에서 처리해야 하니까요.

교통 정보는 하기로 했던 ODsay의 다음과 같은 예외 상황을 모두 고려해야 합니다.
![image](https://user-images.githubusercontent.com/76458724/204124410-cdc829e4-7891-4d34-81d8-7a4aa82beef0.png)
~~자가용과 도보는 T MAP API에서 속도를 받아서 하고 있습니다!~~
~~도보는 사람의 평균 걷기 속도로 하면 되겠지만 자동차는 예외 상황이 많지 않을까요..?~~
자가용과 도보의 속도는 필수가 아니었군요.

저희가 이런 정보까지 전부 파악하기에는 시간 내에는 어려울 거라고 생각합니다...

## 서버 사용
4명 모두 안드로이드 개발자인 상황에서 서버에서 자기장 정보를 주도록 개발하는 것은 어려워 보입니다.
그 다음 대책은 방장 휴대폰인데, 방장 휴대폰의 네트워크와 GPS 상황을 모두 고려해야 하니 어려울 것 같기도 해요.
(이 때는 다음 방장으로 넘기는 방식도 괜찮은 것 같아 보이지만..)
지금 단계에서는 **각 휴대폰에서 자기장을 계산하는 로직**으로 가야 하지 않을까 싶습니다.

**Reply by tnwlsgk1002** [reply]
여기서 문제는 도보 & 자동차 & 대중 교통 평균 속도를 어떻게 구할 것인가 문제여서 캔슬되었습니다...
